File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/mailbox/item_detail/bodies/components/PartyHero.kt
@file:Suppress("PackageNaming", "MagicNumber", "FunctionNaming", "LongMethod")

package app.pantopus.android.ui.screens.mailbox.item_detail.bodies.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.data.api.models.mailbox.v2.PartyDetailDto
import app.pantopus.android.data.api.models.mailbox.v2.PartyRsvpStatus
import app.pantopus.android.ui.components.ConfettiSpray
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * A17.9 — Festive event hero. Three-layer card:
 *  1. Pink-gradient band carrying the trust + party-invite chips, the
 *     time-ago stamp, the "You're invited by …" eyebrow, and the serif
 *     event title. A `ConfettiSpray` overlay is anchored to the top-left.
 *  2. A `PartyDateTile` + location row that overlaps the band by 46 dp.
 *  3. Optional green confirmation banner in the going state.
 */
@Composable
fun PartyHero(
    party: PartyDetailDto,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.categoryParty.copy(alpha = 0.25f), RoundedCornerShape(Radii.lg))
                .testTag("partyHero"),
    ) {
        FestiveBand(party)
        DatePanel(
            party,
            modifier =
                Modifier
                    .padding(horizontal = Spacing.s3)
                    .offset(y = (-46).dp),
        )
        if (party.rsvp == PartyRsvpStatus.Going) {
            GoingBanner(
                party,
                modifier =
                    Modifier
                        .padding(horizontal = Spacing.s3)
                        .padding(bottom = Spacing.s3)
                        .offset(y = (-46).dp),
            )
        } else {
            Spacer(modifier = Modifier.size(Spacing.s0))
        }
    }
}

@Composable
private fun FestiveBand(party: PartyDetailDto) {
    Box(
        modifier =
            Modifier
                .fillMaxWidth()
                .background(
                    Brush.linearGradient(
                        0.0f to PantopusColors.errorBg,
                        0.55f to PantopusColors.categoryParty.copy(alpha = 0.20f),
                        1.0f to PantopusColors.categoryParty.copy(alpha = 0.32f),
                    ),
                ),
    ) {
        ConfettiSpray(modifier = Modifier.size(width = 200.dp, height = 140.dp))

        Column(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = Spacing.s4)
                    .padding(top = Spacing.s3, bottom = 60.dp),
            verticalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Spacing.s1)) {
                TrustPill()
                PartyInvitePill()
                Spacer(modifier = Modifier.weight(1f))
                Text(
                    text = party.timeAgoLabel,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.categoryParty.copy(alpha = 0.7f),
                )
            }
            val firstName = party.host.name.split(" ").firstOrNull() ?: party.host.name
            Text(
                text = "You're invited by $firstName",
                fontSize = 11.sp,
                fontWeight = FontWeight.Black,
                letterSpacing = 0.8.sp,
                color = PantopusColors.categoryParty.copy(alpha = 0.85f),
            )
            val title =
                if (party.rsvp == PartyRsvpStatus.Going) {
                    val month = party.event.date.monthLabel.lowercase().replaceFirstChar { it.uppercase() }
                    "You're going on $month ${party.event.date.dayNumber}"
                } else {
                    party.event.what
                }
            Text(
                modifier = Modifier.semantics { heading() },
                text = title,
                fontSize = 22.sp,
                fontWeight = FontWeight.Black,
                fontFamily = FontFamily.Serif,
                letterSpacing = (-0.3).sp,
                color = PantopusColors.appText,
                lineHeight = 27.sp,
            )
        }
    }
}

@Composable
private fun DatePanel(
    party: PartyDetailDto,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(14.dp))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(14.dp))
                .padding(Spacing.s3),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        PartyDateTile(
            monthLabel = party.event.date.monthLabel,
            dayNumber = party.event.date.dayNumber,
            dayLabel = party.event.date.dayLabel,
        )
        Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(Spacing.s1)) {
            Text(
                text = "${party.event.date.weekday} · ${party.event.date.timeRange}",
                fontSize = 13.5.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appText,
            )
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(5.dp)) {
                PantopusIconImage(
                    icon = PantopusIcon.MapPin,
                    contentDescription = null,
                    size = 12.dp,
                    tint = PantopusColors.categoryParty,
                )
                Text(
                    text = party.event.location,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.appTextStrong,
                )
            }
            Text(
                modifier = Modifier.padding(start = 17.dp),
                text = party.event.locationNote,
                fontSize = 11.sp,
                color = PantopusColors.appTextSecondary,
            )
        }
    }
}

@Composable
private fun GoingBanner(
    party: PartyDetailDto,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(10.dp))
                .background(PantopusColors.successBg)
                .border(1.dp, PantopusColors.successLight, RoundedCornerShape(10.dp))
                .padding(horizontal = Spacing.s2, vertical = 9.dp)
                .testTag("partyHero_goingBanner"),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Box(
            modifier = Modifier.size(20.dp).clip(CircleShape).background(PantopusColors.success),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(
                icon = PantopusIcon.Check,
                contentDescription = null,
                size = 12.dp,
                tint = PantopusColors.appTextInverse,
            )
        }
        Text(
            text =
                buildAnnotatedString {
                    withStyle(SpanStyle(fontWeight = FontWeight.Black)) {
                        append("You're going")
                    }
                    if (party.plusOneCount > 0) {
                        withStyle(SpanStyle(fontWeight = FontWeight.Black)) {
                            append(" · +${party.plusOneCount}")
                        }
                    }
                    party.rsvpConfirmedAtLabel?.let { stamp ->
                        withStyle(
                            SpanStyle(
                                fontWeight = FontWeight.SemiBold,
                                color = PantopusColors.success.copy(alpha = 0.85f),
                            ),
                        ) { append(" · RSVP'd $stamp") }
                    }
                },
            fontSize = 12.sp,
            color = PantopusColors.success,
        )
    }
}

@Composable
private fun TrustPill() {
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(PantopusColors.successBg)
                .padding(horizontal = Spacing.s2, vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(3.dp),
    ) {
        PantopusIconImage(
            icon = PantopusIcon.ShieldCheck,
            contentDescription = null,
            size = 10.dp,
            tint = PantopusColors.success,
        )
        Text(
            text = "Verified",
            fontSize = 10.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.success,
        )
    }
}

@Composable
private fun PartyInvitePill() {
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.categoryParty.copy(alpha = 0.45f), RoundedCornerShape(Radii.pill))
                .padding(horizontal = Spacing.s2, vertical = 3.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        Box(modifier = Modifier.size(6.dp).clip(CircleShape).background(PantopusColors.categoryParty))
        Text(
            text = "Party invite",
            fontSize = 10.sp,
            fontWeight = FontWeight.Black,
            letterSpacing = 0.2.sp,
            color = PantopusColors.categoryParty,
        )
    }
}

File: frontend/apps/ios/Pantopus/Features/Mailbox/MailDetail/Variants/Components/PartyHero.swift
//
//  PartyHero.swift
//  Pantopus
//
//  A17.9 — Festive event hero. Three-layer card:
//    1. Pink-gradient band carrying the trust + party-invite chips, the
//       time-ago stamp, the "You're invited by …" eyebrow, and the
//       serif event title. The `ConfettiSpray` primitive overlays the
//       band — gentle drift in the rose / amber / sky / magic palette
//       lifted from `party.jsx` Confetti.
//    2. A `DateTile` + location row that overlaps the band by 46pt so
//       the panel sits half on / half off the gradient (the design's
//       signature "ticket stub" trick).
//    3. Optional going-state confirmation banner — green check pill,
//       headcount, +1 chip, and the local-time stamp.
//

import SwiftUI

@MainActor
struct PartyHero: View {
    let party: PartyDetailDTO

    var body: some View {
        VStack(spacing: Spacing.s0) {
            festiveBand
            datePanel
                .padding(.horizontal, Spacing.s3)
                .padding(.top, -46)
                .padding(.bottom, party.rsvp == .going ? Spacing.s2 : Spacing.s3)
            if party.rsvp == .going {
                goingBanner
                    .padding(.horizontal, Spacing.s3)
                    .padding(.bottom, Spacing.s3)
            }
        }
        .background(Theme.Color.appSurface)
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg)
                .stroke(Theme.Color.categoryParty.opacity(0.25), lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.lg))
        .shadow(color: Theme.Color.categoryParty.opacity(0.10), radius: 10, x: 0, y: 4)
        .accessibilityIdentifier("partyHero")
    }

    // MARK: - Band

    private var festiveBand: some View {
        ZStack(alignment: .bottomLeading) {
            LinearGradient(
                colors: [
                    Theme.Color.errorBg,
                    Theme.Color.categoryParty.opacity(0.20),
                    Theme.Color.categoryParty.opacity(0.32)
                ],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )

            ConfettiSpray(seed: 24, dotCount: 48)
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
                .allowsHitTesting(false)

            VStack(alignment: .leading, spacing: Spacing.s2) {
                HStack(spacing: Spacing.s1) {
                    PartyHeroTrustPill()
                    PartyInvitePill()
                    Spacer()
                    Text(party.timeAgoLabel)
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(Theme.Color.categoryParty)
                        .opacity(0.7)
                }
                Text("You're invited by \(party.host.name.split(separator: " ").first.map(String.init) ?? party.host.name)")
                    .font(.system(size: 11, weight: .heavy))
                    .tracking(0.8)
                    .foregroundStyle(Theme.Color.categoryParty.opacity(0.85))
                Text(party.rsvp == .going
                    ? "You're going on \(party.event.date.monthLabel.capitalized) \(party.event.date.dayNumber)"
                    : party.event.what
                )
                .font(.system(size: 22, weight: .heavy, design: .serif))
                .tracking(-0.3)
                .foregroundStyle(Theme.Color.appText)
                .fixedSize(horizontal: false, vertical: true)
            }
            .padding(.horizontal, Spacing.s4)
            .padding(.top, Spacing.s3)
            .padding(.bottom, 60)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    // MARK: - Date / location panel

    private var datePanel: some View {
        HStack(spacing: Spacing.s3) {
            DateTile(
                monthLabel: party.event.date.monthLabel,
                dayNumber: party.event.date.dayNumber,
                dayLabel: party.event.date.dayLabel
            )
            VStack(alignment: .leading, spacing: Spacing.s1) {
                Text("\(party.event.date.weekday) · \(party.event.date.timeRange)")
                    .font(.system(size: 13.5, weight: .bold))
                    .foregroundStyle(Theme.Color.appText)
                    .tracking(-0.1)
                HStack(spacing: 5) {
                    Icon(.mapPin, size: 12, color: Theme.Color.categoryParty)
                    Text(party.event.location)
                        .font(.system(size: 12, weight: .semibold))
                        .foregroundStyle(Theme.Color.appTextStrong)
                }
                Text(party.event.locationNote)
                    .font(.system(size: 11))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .padding(.leading, 17)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
        }
        .padding(Spacing.s3)
        .background(Theme.Color.appSurface)
        .overlay(
            RoundedRectangle(cornerRadius: 14)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: 14))
        .shadow(color: Color.black.opacity(0.06), radius: 6, x: 0, y: 4)
    }

    // MARK: - Going banner

    private var goingBanner: some View {
        HStack(spacing: Spacing.s2) {
            ZStack {
                Circle().fill(Theme.Color.success).frame(width: 20, height: 20)
                Icon(.check, size: 12, color: Theme.Color.appTextInverse)
            }
            (
                Text("You're going")
                    .font(.system(size: 12, weight: .heavy))
                    + Text(party.plusOneCount > 0 ? " · +\(party.plusOneCount)" : "")
                    .font(.system(size: 12, weight: .heavy))
                    + Text(party.rsvpConfirmedAtLabel.map { " · RSVP'd \($0)" } ?? "")
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundColor(Theme.Color.success.opacity(0.85))
            )
            .foregroundColor(Theme.Color.success)
            Spacer(minLength: Spacing.s0)
        }
        .padding(.horizontal, Spacing.s2)
        .padding(.vertical, 9)
        .background(Theme.Color.successBg)
        .overlay(
            RoundedRectangle(cornerRadius: 10)
                .stroke(Theme.Color.successLight, lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: 10))
        .accessibilityIdentifier("partyHero_goingBanner")
    }
}

// MARK: - Chips

private struct PartyHeroTrustPill: View {
    var body: some View {
        HStack(spacing: 3) {
            Icon(.shieldCheck, size: 10, color: Theme.Color.success)
            Text("Verified")
                .font(.system(size: 10, weight: .bold))
                .foregroundStyle(Theme.Color.success)
        }
        .padding(.horizontal, Spacing.s2)
        .padding(.vertical, 3)
        .background(Theme.Color.successBg)
        .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
    }
}

private struct PartyInvitePill: View {
    var body: some View {
        HStack(spacing: Spacing.s1) {
            Circle().fill(Theme.Color.categoryParty).frame(width: 6, height: 6)
            Text("Party invite")
                .font(.system(size: 10, weight: .heavy))
                .tracking(0.2)
                .foregroundStyle(Theme.Color.categoryParty)
        }
        .padding(.horizontal, Spacing.s2)
        .padding(.vertical, 3)
        .background(Theme.Color.appSurface)
        .overlay(
            RoundedRectangle(cornerRadius: Radii.pill)
                .stroke(Theme.Color.categoryParty.opacity(0.45), lineWidth: 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.pill))
    }
}

#Preview("Open") {
    PartyHero(party: MailItemSampleData.partyInvite)
        .padding(Spacing.s4)
        .background(Theme.Color.appBg)
}

#Preview("Going") {
    PartyHero(party: MailItemSampleData.partyInviteGoing)
        .padding(Spacing.s4)
        .background(Theme.Color.appBg)
}
